datadog_lambda/tracing.py

```diff
@@ -685,9 +685,32 @@ def create_inferred_span(
     return None
 
 
+def get_service_mapping(service_name):
+    service_mapping = os.getenv("DD_SERVICE_MAPPING", "")
+    mapping = {}
+
+    for entry in service_mapping.split(","):
+        parts = entry.split(":")
+        if len(parts) == 2:
+            key, value = parts
+            mapping[key.strip()] = value.strip()
+
+    return mapping.get(service_name)
+
+
 def create_inferred_span_from_lambda_function_url_event(event, context):
     request_context = event.get("requestContext")
+    api_id = request_context.get("apiId")
     domain = request_context.get("domainName")
+    # Attempt to get the service mapping for the domain so that we don't remap all apigw services
```

Comment:
Hmm, this complexity is necessary but I wonder if we can formalize it a bit more?

Right now a bug could be inadvertently introduced in line 707 or 710, which would `fail through` to the base case of `domain`.

That's good behavior, but when we see logic on different lines our minds don't always connect them. Would a switch statement help make the causality clear? Wdyt?

Reply:
perhaps 
```
api_id = request_context.get("apiId")
domain = request_context.get("domainName")

# Attempt to get the service mapping for the domain so that we don't remap all apigw services
# Allows the customer to have more fine-grained control
apiid_mapping = get_service_mapping(str(api_id))

lambda_url_mapping = None
if not apiid_mapping:
    # If the domain mapping is not found,
    # attempt to get the service mapping for 'lambda_api_gateway'
    lambda_url_mapping = get_service_mapping("lambda_url")

# If neither mapping is found, default to the domain name
if apiid_mapping:
    service_name = apiid_mapping
elif lambda_url_mapping:
    service_name = lambda_url_mapping
else:
    service_name = domain
```

Reply:
```
api_id = request_context.get("apiId")
domain = request_context.get("domainName")

# Attempt to get the service mapping for the domain so that we don't remap all apigw services
# Allows the customer to have more fine-grained control
apiid_mapping = get_service_mapping(str(api_id))

# Attempt to get the service mapping for 'lambda_api_gateway'
lambda_url_mapping = get_service_mapping("lambda_url")

if apiid_mapping:
    service_name = apiid_mapping
elif lambda_url_mapping:
    service_name = lambda_url_mapping
else:
    # If neither mapping is found, default to the domain name
    service_name = domain

```

Reply:
Okay good start - definitely clearer. Let’s define the requirements in plain english real quickly:
Highest priority - user overrides the specific function URL.
Medium priority - user overrides lambda_url
Lowest priority - no customization

To me, this means we should have a pretty clean conditional, and we can avoid mutating a variable unnecessary by refactoring this to its own function (don't actually do this, but for example):
```python
def function_url_mapping(api_id_string):
  if get_service_mapping(api_id_string):
    return get_service_mapping(api_id)
  else if get_service_mapping("lambda_url"):
    return get_service_mapping("lambda_url")
   else:
     return domain
```

But this has some problems. Firstly, `get_service_mapping` re-parses the ENV var with every call, which we only need to do once when the function starts. We don't need a function for this, we could simply make python do this when the class loads. This means refactoring things a bit. I suggested above that we simply use a dict to track the mapping, which now lets us make our work more pythonic.

```python
def function_url_mapping(api_id_string):
  if api_id_string in service_mapping:
    return service_mapping.get(api_id_string)
  else if "lambda_url" in service_mapping:
    return service_mapping.get("lambda_url")
  else:
    return domain
```

How does that read to you? Does that feel more like the business rules we laid out above?

Reply:
100% this looks awesome! I like the added touch of parsing the env var as the class loads, makes it really clean

Reply:
this becomes hard to test though as DD_SERVICE_MAPPING will only be parsed once, when the module loads